### PR TITLE
Feature/driver init from serialbox

### PR DESCRIPTION
**Pull request by elynnwu**
## Purpose

In this PR, we add the capability of initializing from serialbox in driver. The capability of using serialized grid is also added.

## Code changes:

- Dockerfile: include serialbox in PYTHONPATH to avoid having to append path
- driver/run.py: 
    - `DriverState` now takes in `grid_data`, `damping_coefficients`, and `driver_grid_data`
    - `SerialboxConfig` implemented: an example yaml is included, where a namelist override can be passed in `dycore_config` and `physics_config`. It is also necessary to pass in nx_tile, nz, dt_atmos, layout, and time information in the yaml file. 
- Move `TranslateUpdateDWindsPhys` to stencils so it can be imported in driver
- pace/util/namelist: Allow date/time information from fortran namelist so that `SerialboxConfig` can use it
- translate_physics: `edge_vect_s/n` mask updated

## Infrastructure changes:

- test_driver: a new driver test that initializes from serialbox, this runs as part of `test_driver` on jenkins
    - note that we currently use `fv3gfs-physics` default data but could switch to driver specific data once that's available 

## Checklist
Before submitting this PR, please make sure:

- [ ] You have followed the coding standards guidelines established at [Code Review Checklist](https://paper.dropbox.com/doc/Code-Review-Checklist--BD7zigBMAhMZAPkeNENeuU2UAg-IlsYffZgTwyKEylty7NhY).
- [ ] Docstrings and type hints are added to new and updated routines, as appropriate
- [ ] Unit tests are added or updated for non-stencil code changes

### CODE REVIEW

File: driver/pace/driver/run.py
Comment (mcgibbon) [inline driver/pace/driver/run.py:198]: It's unfortunate that we need a stencil factory to do something that involves creating no stencils, but it looks like that is in fact the case because the translate infrastructure uses low-level configuration from StencilFactory in combination with `make_storage_from_shape` instead of using `QuantityFactory`.

I think though that we only need the `backend` value. You should be able to remove the other configuration options above (rebuild, validate_args, format_source, device_sync) and provide a nonsensical value for them (None, probably) that would cause errors if we ever did create stencils through this pathway. This would reduce the dataclass API above to only what's needed.

File: driver/pace/driver/run.py
Comment (mcgibbon) [inline driver/pace/driver/run.py:229]: Can these be private methods?

File: driver/pace/driver/run.py
Comment (mcgibbon) [inline driver/pace/driver/run.py:301]: You can delete this method and call the code below directly where you'd call this function, since the call signatures are similar size and the code contained here would be more explicit.

File: driver/pace/driver/run.py
Comment (mcgibbon) [inline driver/pace/driver/run.py:160]: Can these be private methods/attributes? They should only be called from inside this class.

File: driver/pace/driver/run.py
Comment (mcgibbon) [inline driver/pace/driver/run.py:416]: Are these lines necessary? My understanding was that dacite does this automatically, but I could be wrong.
```suggestion
```

File: driver/pace/driver/run.py
Comment (mcgibbon) [inline driver/pace/driver/run.py:427]: This return statement here makes it a lot harder to tell what this function does. There's code below that does necessary processing to integrate the top level driver configuration with values in DycoreConfig and PhysicsConfig. Either that code is duplicated somehow in `set_kwargs_from_serialbox`, which makes it harder to maintain that code (has to be updated in two places), or that code is not duplicated and the behavior of the rest of the configuration depends on how the initialization is set.

It also looks like right now the value of `initialization_type` makes the non-initialization part of the driver configuration yaml behave differently in ways I wouldn't expect as a user.

Is there any way this can be reduced to `initialization_class=SerialboxConfig` with all the concerns met by `set_kwargs_from_serialbox` treated automatically within the `SerialboxConfig` class? You have full control over the attributes of SerialboxConfig to make it work when you call `dacite.from_dict`, and you have access to a `__post_init__` dataclass method to add/modify arbitrary attributes afterwards.

**Review by mcgibbon — CHANGES_REQUESTED** [review]

File: driver/examples/configs/baroclinic_c12_from_serialbox.yaml
Comment (elynnwu) [reply driver/examples/configs/baroclinic_c12_from_serialbox.yaml:1]: Do we not want to have this in driver/Makefile at all? I thought that we would want to have this in examples because while serialbox may not be a desired option for users, the restart config is expected to be used commonly. So shouldn't we have all the examples?
I think documentation on how to use this yaml file is a good idea (will add this), not just externally but also nice to see easily what options we support and how to use it.

File: driver/Makefile
Comment (elynnwu) [reply driver/Makefile:9]: Fair enough. Do we expect users to run from make file? Because there're some instructions in driver/README, and they don't reference the makefile either.

File: driver/pace/driver/run.py
Comment (elynnwu) [reply driver/pace/driver/run.py:77]: I think we need to consolidate `GridData` and `DriverGridData`, mostly just rolling whatever in driver grid data to grid data. There're also some additional changes we need to add in fortran grid savepoint, currently the variables in `DriverGridData` are not in that savepoint.

File: driver/pace/driver/run.py
Comment (elynnwu) [reply driver/pace/driver/run.py:138]: Yes this is because we need to make a `stencil_config` to use TranslateUpdateDWindsPhys. When we include these grid variables in the gridinfo savepoint, it should look a bit cleaner.

File: driver/pace/driver/run.py
Comment (elynnwu) [reply driver/pace/driver/run.py:198]: Sounds good, keeping `backend` only here. Again, this should look cleaner when we consolidate `GridData` and `DriverGridData`.

Comment (elynnwu) [reply driver/pace/driver/run.py:198]: Making a jira card for that now

File: driver/pace/driver/run.py
Comment (elynnwu) [reply driver/pace/driver/run.py:229]: Changed to private methods

File: driver/pace/driver/run.py
Comment (elynnwu) [reply driver/pace/driver/run.py:160]: Changed the last two. The namelists are actually called in `_set_kwargs_from_serialbox` inside `DriverConfig`. We override the options to be from fortran namelist there.

File: driver/pace/driver/run.py
Comment (elynnwu) [reply driver/pace/driver/run.py:416]: It is necessary because we're making a SerialboxConfig first before making DriverConfig class, and we need backend from stencil_config.

File: driver/pace/driver/run.py
Comment (rheacangeo) [reply driver/pace/driver/run.py:77]: I thought the direction we wanted to go was distinct DriverGridData, PhysicsGridData, and GridData(which could be named DycoreGridData)? You'd rather have one large bucket that has all 3 and contains irrelevant variables when running dycore only or physics only?

Comment (mcgibbon) [reply driver/pace/driver/run.py:77]: I read what @elynnwu wrote as being that we can have a `GridData` class with `grid_data.dycore`, `grid_data.physics` and `grid_data.<whatever we call the conversion part currently referred to as DriverGridData>`. Keeps the convenience of one place to initialize or access the data at the top level, but also the nice interfaces at lower levels where you only need to create relevant variables.

I'm more open on this one versus the configuration classes though, since grid data can be generated with very few free parameters/config options. One bin can be OK as long as there are very few free parameters needed to initialize it.

File: driver/Makefile
Comment (mcgibbon) [reply driver/Makefile:9]: I don't expect user to run this file soon - the Makefile is here for our development purposes. test_mpi is useful to iterate locally on code without the overhead of compiling/running docker. test_mpi_serialbox could also be useful for certain development, but there's no way to run it other than using `.jenkins/test_driver.sh` since it explicitly depends on being in a docker environment (specifically, having a test_data folder at the filesystem root).

You could get it running automatically through this make target, but I think test_driver.sh should be enough.

File: driver/pace/driver/run.py
Comment (elynnwu) [reply driver/pace/driver/run.py:77]: This is a good point. I hadn't thought about the case where a user just wants to run one component when I made the comment. I think we will have to go through the variables one by one to see which way makes more sense, updating the jira card now.

File: driver/pace/driver/run.py
Comment (elynnwu) [reply driver/pace/driver/run.py]: Okay updated per our discussion. It's still not super clean because I couldn't find a way to do this nicely without adding more inputs in the serialbox config yaml. More comments below.

File: fv3core/fv3core/_config.py
Comment (elynnwu) [inline fv3core/fv3core/_config.py:287]: Is there a way to have post_init return a new class and override the initialization? I tried a few variations of calling class methods but no luck, so ended up settling on explicitly setting each variable.

File: driver/Makefile
Comment (elynnwu) [reply driver/Makefile]: Okay, `test_mpi_serialbox` removed from Makefile and moved the command directly to `test_driver.sh`. Keeping `baroclinic_c12_from_serialbox.yaml` in the examples/configs folder because it needs to be somewhere docker can access. I think it makes sense for the example yaml file to be there even though it points to a path that people not using docker will not have access, information on the path is added in README.

File: driver/examples/configs/README.md
Comment (mcgibbon) [inline driver/examples/configs/README.md:1]: This is good to have for now, in the long term it will be hard to maintain this README separate from the code. If and when we put together documentation, we should try to make it so Sphinx automatically documents these details based on dataclass docstrings.

File: driver/examples/configs/baroclinic_c12_from_serialbox.yaml
Comment (mcgibbon) [reply driver/examples/configs/baroclinic_c12_from_serialbox.yaml:1]: What you have right now (the config is here but not the make target) is good. You could also consider putting the config under `driver/tests/` or some other testing directory.

File: driver/pace/driver/run.py
Comment (mcgibbon) [inline driver/pace/driver/run.py:145]: Can these be private attributes?

File: driver/pace/driver/run.py
Comment (mcgibbon) [inline driver/pace/driver/run.py:250]: Can you pass this as an explicit argument to the private methods that use it, instead of passing it through an object attribute? We don't need this data to persist on `self` between calls, and passing it as an argument makes this clearer.

File: driver/pace/driver/run.py
Comment (mcgibbon) [inline driver/pace/driver/run.py:420]: Is this necessary? We expect dycore_config and physics_config should always be included in the yaml.

File: driver/pace/driver/run.py
Comment (mcgibbon) [inline driver/pace/driver/run.py:439]: This still makes me nervous (changing the behavior of the coupling when initialization_type="serialbox"), a consequence of that choice is this has pretty odd behaviors in the case where initialization_type="serialbox" and namelist_override is not used for one or both of dycore_config and physics_config. We don't plan to use it but it's a valid configuration option presented to the user.

Is it that bad to rely on the yaml for the top-level settings (nx_tile, nz, layout) and expect the user to make sure these correspond with the serialbox directory?

Admittedly this is a lot cleaner and less dangerous than before, so I'll leave it up to you. If you keep it this way, I would suggest adding a check that raises an exception if initialization_type="serialbox" and namelist_override is not given for either of dycore_config and physics_config.

File: fv3core/fv3core/_config.py
Comment (mcgibbon) [reply fv3core/fv3core/_config.py:287]: I expected you'd set each variable here. If you like you can

1) make a new object in here with DynamicalCoreConfig.from_f90nml
2) iterate over the dataclass fields of the object or self (they're the same type) and assign with setattr and getattr, this is OK type-wise because you're assigning from and to the same type without adding new attributes.

File: fv3gfs-physics/fv3gfs/physics/_config.py
Comment (mcgibbon) [inline fv3gfs-physics/fv3gfs/physics/_config.py:105]: It would be safer to raise an exception if the override path does not exist.

**Review by mcgibbon — COMMENTED** [review]
Great job! Could be mergeable with few or no changes, not approving yet in case you act on one or more of the change suggestions.

**Review by rheacangeo — APPROVED** [review]
lgtm

File: driver/pace/driver/run.py
Comment (elynnwu) [reply driver/pace/driver/run.py:145]: Yes, forgot to update this now that we don't need to call them from outside the class.

File: driver/pace/driver/run.py
Comment (elynnwu) [reply driver/pace/driver/run.py:250]: Done!

File: driver/pace/driver/run.py
Comment (elynnwu) [reply driver/pace/driver/run.py:420]: It is necessary because we want to use nx_tile, nz, and layout from dycore_config.

File: driver/pace/driver/run.py
Comment (elynnwu) [reply driver/pace/driver/run.py:439]: You can actually still have serialbox initialization and pass in dycore_config and physics_config. I think we don't expect people to do that (assuming if you have serialized data, you ran fortran with a nml file), but in theory, it is allowed and shouldn't break the code.

File: fv3core/fv3core/_config.py
Comment (elynnwu) [reply fv3core/fv3core/_config.py:287]: Great suggestion, I knew there was a better way to utilize `from_f90nml`! Updating this.

File: driver/pace/driver/run.py
Comment (mcgibbon) [reply driver/pace/driver/run.py:447]: What I mean is, the normal expectation here is that you don't pass `npx` in `dycore_config`, and instead you pass `nx_tile`. If you take a configuration file that does this and modify the initailization_type, npx will be set to 0 and no exception will be raised. This could be pretty hard to detect and debug.

Comment (mcgibbon) [reply driver/pace/driver/run.py:447]: Overriding the owning class behavior based on the choice of subclass for initialization makes me nervous because of how easy it is to not notice potential bugs like that, and how hard it is to detect them when they do happen.

If we consider the "serialbox" initialization to be a use-at-your-own-risk configuration that may cause strange bugs then this is OK, but I'm not sure the tradeoff of not copying these variables from the namelist to the yaml outweighs the risk of us running into one of these bugs. We can merge this as-is if you think it will be fine.

Comment (mcgibbon) [reply driver/pace/driver/run.py:447]: Actually, is there a more important bug here? Does `dt_atmos` as well as `days`, `hours`, `minutes`, `seconds` need to be copied from DycoreConfig to DriverConfig as well?

**Review by mcgibbon — APPROVED** [review]
LGTM, thanks for all the work avoiding side-effects on this one!

**Comment by elynnwu** [comment]
launch jenkins

**Pull request merged**
